x-pack/plugins/infra/public/containers/logs/log_filter/log_filter_state.ts

```diff
@@ -5,103 +5,172 @@
+
       try {
-        const parsedQuery = parseQuery(filterQuery);
-        setLogFilterState((previousLogFilterState) => ({
-          ...previousLogFilterState,
-          filterQuery: {
-            parsedQuery,
-            serializedQuery: JSON.stringify(parsedQuery),
-            originalQuery: filterQuery,
-          },
-          filterQueryDraft: filterQuery,
-          validationErrors: [],
-        }));
+        parseQuery(query);
       } catch (error) {
+        throw new QueryParsingError(error);
+      }
+    },
+    [parseQuery]
+  );
+
+  const handleValidationError = useCallback(
+    (error: Error) => {
+      if (error instanceof UnsupportedLanguageError) {
+        toasts.addError(error, { title: errorToastTitle });
+        queryString.setQuery(DEFAULT_QUERY);
+      } else if (error instanceof QueryParsingError) {
+        toasts.addError(error, { title: errorToastTitle });
+      }
+    },
+    [queryString, toasts]
+  );
+
+  const [logFilterState, setLogFilterState] = useState<ILogFilterState>(() => {
+    const query = queryString.getQuery();
+    try {
+      validateQuery(query);
+      return {
+        filterQuery: getLogFilterQuery(query as Query),
+        queryStringQuery: query,
+        validationError: null,
+      };
+    } catch (error) {
+      return {
+        filterQuery: null,
+        queryStringQuery: query,
+        validationError: error,
+      };
+    }
+  });
+
+  useEffect(() => {
+    if (logFilterState.validationError) {
+      handleValidationError(logFilterState.validationError);
+    }
+  }, [handleValidationError, logFilterState.validationError]);
+
+  const applyLogFilterQuery = useCallback(
+    (validFilterQuery: Query) => {
+      const logFilterQuery = getLogFilterQuery(validFilterQuery);
+      if (logFilterQuery) {
         setLogFilterState((previousLogFilterState) => ({
           ...previousLogFilterState,
-          validationErrors: [`${error}`],
+          filterQuery: logFilterQuery,
         }));
       }
     },
-    [cancelPendingValidation, parseQuery]
+    [getLogFilterQuery]
+  );
+
+  useSubscription(
+    useMemo(() => {
+      return merge(of(undefined), queryString.getUpdates$()); // NOTE: getUpdates$ uses skip(1) so we do this to ensure an initial emit of a value.
+    }, [queryString]),
+    useMemo(() => {
+      return {
+        next: () => {
+          const query = queryString.getQuery();
+          try {
+            validateQuery(query);
+            setLogFilterState((previousLogFilterState) => ({
+              ...previousLogFilterState,
+              queryStringQuery: query,
+              validationError: null,
+            }));
+            applyLogFilterQuery(query as Query);
+          } catch (error) {
+            setLogFilterState((previousLogFilterState) => ({
+              ...previousLogFilterState,
+              queryStringQuery: query,
+              validationError: error,
+              filterQuery: null,
+            }));
+          }
```

Comment:
I'm having a hard time following the semantics of these functions and their call sequence:

- the called `validateQuery()` calls `parseQuery()` and re-throws a wrapper error as a side-effect
- the called `setLogFilterState()` merges the new query and "no error" into the local state
- the called `applyLogFilterQuery()` calls `getLogFilterQuery()` which also calls `parseQuery()` and merges its result into `filterQuery`

In effect we parse multiple times and throw away the previous query when the newly applied doesn't parse correctly.

How about consolidating `getLogFilterQuery`, `validateQuery` and `applyLogFilterQuery` into a

```typescript
const getNewLogFilterState = (newQueryString: string) => (previousLogFilterState: ILogFilterQuery): ILogFilterQuery => { /* ... */ }
```

which `parseQuery()`s, catches exceptions and returns a new data structure with the proper `filterQuery`, `queryStringQuery` and `validationError`.

This could then be called like

```typescript
getNewLogFilterState(queryString.getQuery())(initialLogFilterState)
```

in the `setState` initializer and

```typescript
setLogFilterState(getNewLogFilterState(queryString.getQuery()))
```

in the subscription's `next` callback without duplication.

Does that make sense?

Reply:
Makes sense. I went through a few different orders and versions of this code. If we're happy with the overall flow now (bar the minor bits about the error and the defaults) I can consolidate this into one function.

Reply:
Thanks :pray: It seems to work as expected except for `filterQuery` being emptied on parsing error. The previous behavior of keeping the last working query and only replacing it with the current draft when it is valid seemed reasonable. It would allow auto-refresh and incremental loading to keep working with the valid previous query.

I don't want to cause unnecessary work here. :innocent: We probably just want to have a readable implementation we can base our upcoming state refactoring on.